File: descripe/imapp_modisl2/src/leocat/src/MOD_PRAlg29/get_regstd.c
/*
!F77

!DESCRIPTION:
  Computes standard deviation of (center) pixel of interest and the     
  surrounding eight pixel values for indicated channel.   


!Input parameters:
   band          MODIS band number (1-36)

!Output Parameters:
   sigma         Standard deviation of indicated MODIS channel
 
   Other outputs stored in structure variable 'rg_var' of type "regional_var"
   in pixel.h (rg_var.mean).
    

!Revision History:
   10/04 Collection 5b   R. Frey
   Added calculation and output of mean value.
   
   11/07 Converted to C  R. Frey

!Team-unique Header:

!References and Credits:
   See Cloud Mask ATBD.

!END
***********************************************************************/

/* Includes */

#include <stdio.h>
#include <math.h>
#include "pixel.h"
#include "mask_processing_constants.h"
#include "granule.h"


float get_regstd(int band)


{

/**********************************************************************/
	
/*  Declarations */
	
	int imvl, imve, nl, ne;
	int i,k;
	
	long int j;
	long int ide;
	
	float sigma;
	
	double n, sum, sumsq, tbb;
	double a, sqsum, num, den, sig, mn;

/**********************************************************************/	
	
/*  Initializations */
 
    n = 0.0;
    sum = 0.0;
    sumsq = 0.0;
    sqsum = 0.0;
    num = 0.0;
    den = 0.0;
    
/**********************************************************************/    

    if(pxin.line_edge && pxin.scan == 0) {
      nl = 2;
      imvl = 0;
    }
    else if(pxin.line_edge && pxin.scan > 0) {
      nl = 2;
      imvl = 1;
    }
    else {
      nl = nlcntx;
      imvl = 1;
    }
    
    if(pxin.elem_edge && pxin.elem == 0) {
      ne = 2;
      imve = 0;
    }
    else if(pxin.elem_edge && pxin.elem > 0) {
      ne = 2;
      imve = 1;
    }
    else {
      ne = necntx;
      imve = 1;
    }

//  Get standard deviation about central pixel.
    
    for(i=0; i<nl; ++i) {
      for(k=0; k<ne; ++k) {
    	
        j = ( (pxin.scan + (i - imvl)) * neles) + pxin.elem;
        ide = j + (k - imve);  
	    tbb = (double)(*( ( g_rad[band-1] ) + ide ));
//      printf("tbb: %ld %ld %f \n", j, ide, tbb);
	    
	    if( rintf(tbb) != rintf(bad_data) ) {
	      n = n + (double)(1.0);
	      sum = sum + tbb;          
	      sumsq = sumsq + (tbb * tbb);
	    }
          
      }
    } 

    if(n > 1.0) {
      sqsum = sum * sum;
      num = (n * sumsq) - sqsum;
      den = n * (n - (double)(1.0));
      a = num / den;
      sig = sqrt(a);
      mn = sum / n;
    }
    else if(n == 1.0) {
      sig = 0.0;
      mn = sum;
    }
    else {
      sig = (double)bad_data;
      mn = (double)bad_data;
    } 
    sigma = (float)sig;
    rg_var.mean = (float)mn;
    
//  printf("std dev: %d %d %d %d %d %d \n", pxin.line_edge, pxin.elem_edge, imvl, imve, ne, nl);
//  printf("std dev: %f %f %f %f %f %f %f %f \n", n, sum, sumsq, sqsum, num, den, sig, mn);

    return sigma;
        
}

File: descripe/imapp_modisl2/src/cloudmask/src/get_regstd.f
      subroutine get_regstd(indat,kele,maxele,line_edge,klin,band,
     *                      sigma,mean)

      implicit none
      save

c---------------------------------------------------------------------
c!F77
c
c!DESCRIPTION:
c     Computes standard deviation of (center) pixel of interest and the     
c     surrounding eight pixel values for indicated channel.   
c
c
c!Input parameters:
c indat         Array containing nlcltx number of lines of
c               reflectance or BT values for each channel
c kele          First element pixel in the region (context)
c maxele        Maximum number of pixels in a scan
c line_edge     Logical variable indicating first or last line of
c               data in a granule
c klin          Counter indicating number of lines processed
c band          MODIS band number (1-36)
c
c!Output Parameters:
c sigma         Standard deviation of indicated MODIS channel
c mean          Mean of pixel reflectances or brightness temperatures
c
c!Revision History:
c 10/04 Collection 5b   R. Frey
c Added calculation and output of mean value.
c
c!Team-unique Header:
c
c!References and Credits:
c See Cloud Mask ATBD-MOD35.
c
c!END
c---------------------------------------------------------------------------

      INCLUDE 'global.inc'

c     scalar arguments
      integer band,kele,maxele,klin
      real sigma,mean
      logical line_edge

c     array arguments
      real indat(npixel,nlcntx,inband)

c     local scalars
      integer imv,ide,i,j,k,nl,debug,h_output
      double precision n,num,den,sqsum,sumsq,sum,tbb,sig,mn

c     local arrays
      real rgdata(nlcntx,necntx)

c ... Common statement for debug purposes
      common / bug / debug, h_output

c ... debug statement ............................................
      if (debug .gt. 0) then
        write(h_output,'(10x/,''Within routine get_regstd.f'',/)')
      endif
c ................................................................

c     Initialize variables.

      imv = 0
      ide = 0
      do 10 i = 1 , nlcntx
        do 20 j = 1 , necntx
          rgdata(i,j) = 0.0
   20   continue
   10 continue

c ... put pixel data into context data array.

      imv = ((necntx-1)/2) + 1

      if(line_edge .and. klin .eq. 1) then
        nl = 1
      else if(line_edge .and. klin .gt. 1) then
        nl = 2
      else
        nl = nlcntx
      end if

      do 100 i = 1,nl
        do 200 j = 1,necntx
          ide = kele + (j-imv)
          if(ide .gt. 0 .and. ide .le. maxele) then
            if (indat(ide,i,band).gt.0.0 .and.
     +              indat(ide,i,band).lt.1000000.0) then
              rgdata(i,j) = indat(ide,i,band)
            else
              rgdata(i,j) = bad_data
            end if
          else
            rgdata(i,j) = bad_data
          end if
  200   continue
  100 continue

c     Get standard deviation central pixel.

      n = 0.d0
      sum = 0.d0
      sumsq = 0.d0
      do 400 i = 1,nl
        do 500 j = 1,necntx
	  tbb = dble(rgdata(i,j))
	  if(abs(tbb-bad_data) .gt. 0.1) then
	    n = n + 1.d0
	    sum = sum + tbb          
	    sumsq = sumsq + (tbb * tbb)
          end if
 500    continue
 400  continue 

      if(n .gt. 1.d0) then
        sqsum = sum * sum
        num = (n * sumsq) - sqsum
        den = n * (n-1)
        sig = dsqrt(num / den)
        mn = sum / n
      else if(n .eq. 1.d0) then
        sig = 0.d0
        mn = sum
      else
        sig = dble(bad_data)
        mn = dble(bad_data)
      end if 
      sigma = sngl(sig)
      mean = sngl(mn)

c ....debug statement ............................................
      if (debug .gt. 0) then
        write(h_output,'(3f12.4)') rgdata
        write(h_output,'(10x,'' Standard dev. calc. '',/,l2,5i6,8f12.4/)')
     *    line_edge,klin,nl,kele,maxele,band,sqsum,sumsq,n,num,den,
     *    sig,sigma,mean
      endif
c ................................................................

      return
      end
